import-scripts/clear-persistence-caches-for-changed-studies.sh

```diff
@@ -0,0 +1,450 @@
+    return $?
+}
+
+# looks for pending work which is available and sets globals selected_cache_reset_work_task_db and selected_cache_reset_work_task_trigger_filepath if anything is found
+# cache reset tasks are preferred to study reset tasks
+# usage : select_cache_reset_work_task_trigger db_label
+function select_cache_reset_work_task_trigger() {
+    # iterate through work task trigger files - skip any locked (in progress) tasks. completed task triggers should already have been cleared, so assume none are complete but catch race failure later in perform_selected_cache_reset_task
+    # first check if any database reset task is pending
+    for portal_db in $PORTAL_DB_LIST ; do
+        selected_cache_reset_work_task_db="$portal_db"
+        db_cache_reset_filepath="$TMP_DIR_ROOT/$portal_db/$CACHE_RESET_NEEDED_FILENAME"
+        if flock_available_for_filepath "$db_cache_reset_filepath" ; then
+            selected_cache_reset_work_task_trigger_filepath="$db_cache_reset_filepath"
+            return 0
+        fi
+    done
+    # second check if any study reset task is pending
+    for portal_db in $PORTAL_DB_LIST ; do
+        selected_cache_reset_work_task_db="$portal_db"
+        tmp_study_subdir="$TMP_DIR_ROOT/$portal_db"
+        for filepath in $tmp_study_subdir/*.$STUDY_CACHE_INVALIDATION_NEEDED_SUFFIX ; do
+            if [ -f "$filepath" ] ; then
+                if flock_available_for_filepath "$filepath" ; then
+                    selected_cache_reset_work_task_trigger_filepath="$filepath"
+                    return 0
+                fi
+            fi
+        done
+    done
+    # nothing is pending
+    return 1
+}
+
+# examines database tmp subdir for presence of a matching succeeded status file, given a database label and a trigger filepath
+# usage : trigger_filepath_is_marked_as_succeeded db_label trigger_filepath
+function trigger_filepath_is_marked_as_succeeded() {
+    db="$1"
+    trigger_filepath="$2"
+    tmp_study_subdir="$TMP_DIR_ROOT/$db"
+    filename="$(basename -- $trigger_filepath)"
+    filerootname="${filename%.*}"
+    trigger_filepath="$tmp_study_subdir/$filerootname.$STUDY_CACHE_INVALIDATION_NEEDED_SUFFIX"
+    succeeded_filepath="$tmp_study_subdir/$filerootname.$STUDY_CACHE_INVALIDATION_SUCCEEDED_SUFFIX"
+    if [ -f "$succeeded_filepath" ] ; then
+        return 0 # yes, marked as succeeded
+    fi
+    return 1
+}
+
+# clears the persistence cache, either using the cache api for study cache reset tasks, or the kubectl based script for full database cache clearing tasks
+# proper function depends on having the selected cache reset work task set in global environment variables
+# usage : perform_selected_cache_reset_task
+function perform_selected_cache_reset_task() {
+    # if (unlikely) task has been successfully performed while we selected it, exit
+    if trigger_filepath_is_marked_as_succeeded "$selected_cache_reset_work_task_db" "$selected_cache_reset_work_task_trigger_filepath" ; then
```

Comment:
Is this really needed?

Reply:
Not really. The original thinking was that multiple processes can be running in PHASE TWO and it is hypothetically conceivable that one process might stop at the time between selecting the triggered task to perform and the obtaining of the lock to do the task. However, this case does not need to be covered because:
- after the trigger file is selected, the next step is to lock the file, making an interruption very unlikely
- it is also extremely unlikely that during this momentary vulnerability, another process could acquire the lock, accomplish the entire task, mark successful status and release the lock before this process advanced at all. It is practically certain that the other process would be holding the lock at the time we attempted to get it, so we would fail to obtain the lock and thereby exit.
- even if the envisioned scenario managed to happen, the only effect would be to clear the cache a second time for the trigger, which is not harmful.
Function and function call removed.